models/gpg_key.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+
+	"github.com/go-xorm/xorm"
+	"golang.org/x/crypto/openpgp"
+	"golang.org/x/crypto/openpgp/packet"
+)
+
+// GPGKey represents a GPG key.
+type GPGKey struct {
+	ID                int64     `xorm:"pk autoincr"`
+	OwnerID           int64     `xorm:"INDEX NOT NULL"`
+	KeyID             string    `xorm:"INDEX TEXT NOT NULL"`
+	PrimaryKeyID      string    `xorm:"TEXT"`
+	Content           string    `xorm:"TEXT NOT NULL"`
+	Created           time.Time `xorm:"-"`
+	CreatedUnix       int64
+	Expired           time.Time `xorm:"-"`
+	ExpiredUnix       int64
+	Added             time.Time `xorm:"-"`
+	AddedUnix         int64
+	SubsKey           []*GPGKey `xorm:"-"`
+	Emails            []*EmailAddress
+	CanSign           bool
+	CanEncryptComms   bool
+	CanEncryptStorage bool
+	CanCertify        bool
+}
+
+// BeforeInsert will be invoked by XORM before inserting a record
+func (key *GPGKey) BeforeInsert() {
+	key.AddedUnix = time.Now().Unix()
+	key.ExpiredUnix = key.Expired.Unix()
+	key.CreatedUnix = key.Created.Unix()
+}
+
+// AfterInsert will be invoked by XORM after inserting a record
+func (key *GPGKey) AfterInsert() {
+	log.Debug("AfterInsert Subkeys: %v", key.SubsKey)
+	sess := x.NewSession()
+	defer sessionRelease(sess)
+	sess.Begin()
+	for _, subkey := range key.SubsKey {
+		if err := addGPGKey(sess, subkey); err != nil {
```

Comment:
Yes that was because I started with the same structure has https://github.com/go-gitea/gitea/blob/master/models/ssh_key.go#L396 but all checks on data are made in parseGPGKey so no need for addGPGKey.